<a href="https://colab.research.google.com/github/Lexxxi42/Data-Preprocessing/blob/main/practice-2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Блок 1. NumPy и векторизованная обработка данных

In [ ]:
import numpy as np

## Задание 1.1. Базовые свойства массивов

*Вы анализируете данные о заказах интернет-магазина.*



*   delivery_time - время доставки в минутах;
*   order_price - стоимость заказа в рублях;
*   items_count - количество товаров в заказе.





In [ ]:
delivery_time = np.array([35, 42, 28, 51, 39, 63, 31, 45])
order_price = np.array([1200, 3500, 800, 4200, 1900, 5600, 1500, 2800])
items_count = np.array([2, 5, 1, 7, 3, 8, 2, 4])

Выведем shape, ndim и size для каждого массива

In [ ]:
def get_info(arr):
  return print(f'shape = {arr.shape}\nndim = {arr.ndim}\nsize = {arr.size}\n{'-'*10}\n')

print(f'Delivery_time:')
get_info(delivery_time)

print(f'Order_price:')
get_info(order_price)

print(f'Items_count:')
get_info(items_count)

Delivery_time:
shape = (8,)
ndim = 1
size = 8
----------

Order_price:
shape = (8,)
ndim = 1
size = 8
----------

Items_count:
shape = (8,)
ndim = 1
size = 8
----------



> *Что такое size?*
>
> Общее количество элементов в массиве


Проверим, что наблюдний в массивах одинаковое количество

In [ ]:
len(delivery_time) == len(order_price) == len(items_count) # все три массива одинаковы по количеству наблюдений

True

> *Объясните, почему каждый индекс в трёх массивах должен соответствовать одному и тому же заказу.*
>
> Потому что все свойства объекта в массивах привязаны к индексу, и если эти данные перемешать, то будет невозможно их сопоставить согласно реальным данным, это же не словарь

**Вопрос для анализа**

`Что произойдёт с результатами анализа, если delivery_time содержит 8 наблюдений, а order_price — только 7? Почему это является проблемой не только с точки зрения Python, но и с точки зрения качества данных?`

> Размерность данных окажется различной, и с точки зрения python анализ и операции (совместные) с такими массивами будут падать с ошибкой, тк их размерность не совпадает. С точки зрения качетва данных, если обойти технически ограничения, то появятся дргуие проблемы: полноты, целостности

## Задание 1.2. Двумерные массивы

Объединим данные в единую матрицу

In [ ]:
orders = np.column_stack([delivery_time, order_price, items_count])
orders

array([[  35, 1200,    2],
       [  42, 3500,    5],
       [  28,  800,    1],
       [  51, 4200,    7],
       [  39, 1900,    3],
       [  63, 5600,    8],
       [  31, 1500,    2],
       [  45, 2800,    4]])

Узнаем shape

In [ ]:
orders.shape

(8, 3)

Данные первого заказа

In [ ]:
orders[0]

array([  35, 1200,    2])

Время доставки всех заказов

In [ ]:
orders[:,0].sum()

np.int64(334)

Стоимость всех заказов

In [ ]:
orders[:,1].sum()

np.int64(21500)

Количество товаров в шестом заказе

In [ ]:
orders[5][2]

np.int64(8)

С помощью срезов получим все заказы после третьего

In [ ]:
orders[2:,:2]

array([[  28,  800],
       [  51, 4200],
       [  39, 1900],
       [  63, 5600],
       [  31, 1500],
       [  45, 2800]])

Первые пять заказов и первые два признака

In [ ]:
orders[:5]

array([[  35, 1200,    2],
       [  42, 3500,    5],
       [  28,  800,    1],
       [  51, 4200,    7],
       [  39, 1900,    3]])

In [ ]:
orders[:,:2]

array([[  35, 1200],
       [  42, 3500],
       [  28,  800],
       [  51, 4200],
       [  39, 1900],
       [  63, 5600],
       [  31, 1500],
       [  45, 2800]])

**Вопрос для анализа**

`Объясните, почему:
orders.shape
возвращает (8, 3), а:
orders.size
возвращает 24.
Что характеризует каждый из этих показателей?`

> shape возвращает форму и количество элементов в каждом из измерений, а size количество элементов 8*3=24

## Задание 1.3. View и Copy

Выделим из матрицы orders данные о скорости доставки

In [ ]:
delivery = orders[:, 0]
delivery

array([35, 42, 28, 51, 39, 63, 31, 45])

Изменим первый элемент на 1000

In [ ]:
delivery[0] += 1000
delivery

array([1035,   42,   28,   51,   39,   63,   31,   45])

Проверим orders

In [ ]:
orders

array([[1035, 1200,    2],
       [  42, 3500,    5],
       [  28,  800,    1],
       [  51, 4200,    7],
       [  39, 1900,    3],
       [  63, 5600,    8],
       [  31, 1500,    2],
       [  45, 2800,    4]])

`Объясните, почему изменение delivery повлияло на исходную матрицу.`

> Объекты изменяемые (списки, словари, массивы и тд) передаются по ссылке а не по значению, если не использовать .copy(), то переменные будут указывать на одну и ту же область памяти

Восстановим матрицу orders

In [ ]:
orders = np.column_stack([delivery_time, order_price, items_count])
orders

array([[  35, 1200,    2],
       [  42, 3500,    5],
       [  28,  800,    1],
       [  51, 4200,    7],
       [  39, 1900,    3],
       [  63, 5600,    8],
       [  31, 1500,    2],
       [  45, 2800,    4]])

Создадим независимую копию

In [ ]:
delivery_copy = orders[:, 0].copy()

Изменим на 1000

In [ ]:
delivery_copy[0] += 1000
delivery_copy

array([1035,   42,   28,   51,   39,   63,   31,   45])

Проверим orders

In [ ]:
orders

array([[  35, 1200,    2],
       [  42, 3500,    5],
       [  28,  800,    1],
       [  51, 4200,    7],
       [  39, 1900,    3],
       [  63, 5600,    8],
       [  31, 1500,    2],
       [  45, 2800,    4]])

**Вопрос для анализа**

`Почему понимание различия между view и copy важно при обработке реальных наборов
данных?
Приведите пример ситуации, в которой случайное изменение view может привести к ошибке
в дальнейшем анализе.`

> view, в отличие от copy, не создает новую ячейку памяти, а просто ссылкается на часть данных из большой области памями
>
> Допустим мы берем массив:
>
> y = np.array([1200, 3500, 800, 4200, 1900, 5600, 1500, 2800])
>
> и его срез x = y[0:3] # x = ([1200, 3500, 800])
>
> пытаемся поменять значение в новой переменной x[0] = 0 # x = ([0, 3500, 800])
>
> но меняется и исходный массив # y = ([0, 3500, 800, 4200, 1900, 5600, 1500, 2800])

## Задание 1.4. Фильтрация с помощью булевых масок

In [ ]:
delivery_time_copy = np.array([35, 42, 28, 51, 39, 63, 31, 45]).copy()

Создадим булеву маску для заказов, время доставки которых более 45 минут включительно

In [ ]:
mask = delivery_time_copy >= 45
mask

array([False, False, False,  True, False,  True, False,  True])

Получим проблемные заказы по маске

In [ ]:
critical_delivery_time = delivery_time[mask]
critical_delivery_time

array([51, 63, 45])

Количество заказов

In [ ]:
len(critical_delivery_time)

3

Среднее время доставки

In [ ]:
critical_delivery_time.sum()/len(critical_delivery_time)

np.float64(53.0)

**Вопрос для анализа**

`В чём различие между условиями:
delivery_time > 45
и
delivery_time >= 45
Как изменение граничного условия может повлиять на бизнес-метрику?`

> в первом случае 45 не включается, а во втором включается
>
> это может сильно повлиять на метрики, тк если, напримиер, неравенство нестрогое >=45, то все все заказы доставленные ровно за 45 минут будут записаны в категорию опоздание, хотя в бизнесс логике таковыми не являются

## Задание 1.5. Фильтрация по нескольким условиям

Создадим булеву маску для заказов, стоимость которых более 3000 и с колчеством товаров более 4

In [ ]:
mask = (order_price > 3000) & (items_count > 4)
mask

array([False,  True, False,  True, False,  True, False, False])

Получим матрицу

In [ ]:
richy = orders[mask]
richy

array([[  42, 3500,    5],
       [  51, 4200,    7],
       [  63, 5600,    8]])

Стоимость каждого

In [ ]:
richy_cost = richy[:,1]
richy_cost

array([3500, 4200, 5600])

Время доставки каждого

In [ ]:
richy_time = richy[:,0]
richy_time

array([42, 51, 63])

Среднее время доставки

In [ ]:
richy_time.sum()/len(richy_time)

np.float64(52.0)

**Вопрос для анализа**

`Почему анализ комбинации нескольких признаков может дать более полезный результат для
бизнеса, чем анализ каждого признака отдельно?
Приведите пример ситуации, когда отдельный показатель выглядит нормально, но комбинация
нескольких признаков выявляет проблему.`

> Анализ комбинации нескольких признаков позволяет увидеть контекст и скрытые зависимости, которые невозможно заметить по отдельным метрикам
>
> Допустим среднее время доставки 25 минут. Для бизнеса это идеальный показатель. Средний рейтинг блюд 4.8. Те лиенты в восторге
> Если мы соединим эти данные и посмотрим на заказы, где время доставки составило менее 15 минут, то увидим, что рейтинг там падает до 2, а в комментариях пишут: "Пицца холодная". **По отдельности метрики отличные. Но комбинация выявила сбой**

# Блок 2. Описательная статистика

## Задание 2.1. Влияние выбросов на статистические показатели

Есть массив времени доставки заказов

In [ ]:
delivery_time = np.array([8, 9, 10, 11, 12])

Среднее

In [ ]:
first_mean = np.mean(delivery_time)
first_mean

np.float64(10.0)

Медиана

In [ ]:
first_median = np.median(delivery_time)
first_median

np.float64(10.0)

Добавим еще наблюдение

In [ ]:
delivery_time_corrupted = np.append(delivery_time, 1000)

Среднее

In [ ]:
second_mean = np.mean(delivery_time_corrupted)
second_mean

np.float64(175.0)

Медиана

In [ ]:
second_median = np.median(delivery_time_corrupted)
second_median

np.float64(10.5)

Сравним изменения

In [ ]:
second_mean - first_mean

np.float64(165.0)

In [ ]:
second_median - first_median

np.float64(0.5)

**Вопросы для анализа**

1. Почему среднее так сильно изменилось?
> Среднее арифметическое учитывает все значения в выборке, появилось очень большое значение и сренее выросло за счет выброса
2. Почему медиана изменилась значительно меньше?
> Медиана более усойчива к выбросам, тк беред значение посередине отсортированного массива
3. Какой показатель лучше описывает типичное время доставки при наличии редких
экстремальных значений?
> При наличии сильных выбросов типичное время доставки лучше всего описывает медиана. Среднее значение в таких условиях будет завышенным и создаст ложное впечатление, что время доставки велико хотя это касается единичных заказов
4. Приведите пример, когда среднее значение является более информативным
показателем, чем медиана.
> Среднее значение нужно, когда бизнесу нужно оценить общий объем ресурсов, совокупные затраты или финальный экономический результат.
>
> Например, если 6 дней курьеры доставляли заказы в пределах 20 минут, а на 7 день доставляют за 30-40, то необходимо учитывать среднее для расчета скидки из-за задержки заказа, так как если брать медиану, то ив 7 день большинство заказов будут идти со скидкой

## Задание 2.2. Процентили и SLA

Интернет-магазин заявляет:

> Не менее 90% заказов доставляются за 40 минут или быстрее.
>
> Для проверки SLA смоделируйте 1000 наблюдений времени доставки.
>
> Используйте нормальное распределение со следующими параметрами:

• среднее — 30 минут;

• стандартное отклонение — 5 минут.

In [ ]:
np.random.seed(42)
delivery_simulated = np.random.normal(
 loc=30,
 scale=5,
 size=1000
)

Среднее значение

In [ ]:
delivery_simulated.mean()

np.float64(30.096660279111628)

Стандартное отклонение

In [ ]:
delivery_simulated.std()

np.float64(4.893631038736771)

Медиана

In [ ]:
np.median(delivery_simulated)

np.float64(30.12650306117444)

Рассчитаем процентили

In [ ]:
np.quantile(delivery_simulated, 0.5)

np.float64(30.12650306117444)

In [ ]:
np.quantile(delivery_simulated, 0.9)

np.float64(36.52822600933569)

In [ ]:
np.quantile(delivery_simulated, 0.95)

np.float64(38.38486264207573)

In [ ]:
np.quantile(delivery_simulated, 0.99)

np.float64(41.57984750401471)

**Вопросы для анализа**

1. Что означает p90 для времени доставки?
> 90% всех заказов доставляются быстрее или ровно за это время, и только 10% превышают этот порог
2. Что означает p95?
> 95% заказов выполняются быстрее или ровно за это время, и лишь 5% превышают этот порог
3. Почему для распределения времени доставки p95 обычно выше p90?
> Все самые долгие заказы лежат в более высоких процентилях
4. Что изменится в управлении логистикой, если компания начнёт ориентироваться не только на p90, но и на p95?
> Компании придется работать с более системными и более сложными логистическими проблемами, которые максимально затягивают доставку

**Продуктовый вывод**

Сформулируйте вывод для менеджера продукта в 2–3 предложениях:

1. выполняется ли SLA;
> SLA выполняется для подавляющего большинства, так как медиана удерживается (30 минут). 90% заказов привозят в пределах 36,5 минут (не превышает 40 минут). Те 95% заказов доставляются за 40 минут или быстрее.
2. насколько хорошо контролируется длинный хвост распределения.
Хвост распределения (до р95) не превышает 40 минут. И только 1% превышает 41,6 минуты

## Задание 2.3. Разброс и предсказуемость

Две курьерские службы выполнили по пять одинаковых доставок.

In [ ]:
service_A = np.array([28, 29, 30, 31, 32])
service_B = np.array([15, 20, 30, 40, 45])

Среднее

In [ ]:
service_A.mean()

np.float64(30.0)

In [ ]:
service_B.mean()

np.float64(30.0)

Медиана

In [ ]:
np.median(service_A)

np.float64(30.0)

In [ ]:
np.median(service_B)

np.float64(30.0)

Стандартное отклонение

In [ ]:
service_A.std()

np.float64(1.4142135623730951)

In [ ]:
service_B.std()

np.float64(11.40175425099138)

90-й процентиль

In [ ]:
np.quantile(service_A, 0.9)

np.float64(31.6)

In [ ]:
np.quantile(service_B, 0.9)

np.float64(43.0)

**Вопросы для анализа**

1. Почему одинаковые или близкие средние значения не означают одинаковое качество работы?
> Среднее значение показывает лишь общую сумму времени, деленную на количество заказов, при этом игнорируя разброс данных. Служба А работает стабильно, а службу Б постоянно мотает от 15 минут до 45
2. Какая служба является более предсказуемой?
> Служба А
3. Какую службу вы выбрали бы для доставки дорогой электроники?
> Служба А
4. Какие статистические показатели повлияли бы на ваше решение?
> СКО и p90

## Задание 2.4. Корреляция

Предположим, что увеличение количества товаров в заказе может приводить к увеличению времени его обработки и доставки.

In [ ]:
np.random.seed(42)
items = np.random.randint(1, 10, 100)
delivery_time_correlated = (
 20
 + 3 * items
 + np.random.normal(0, 3, 100)
)

Рассчитаем коэффициент корреляции Пирсона между items и delivery_time_correlated

In [ ]:
corr_matrix = np.corrcoef(items, delivery_time_correlated)
corr_matrix

array([[1.        , 0.93432937],
       [0.93432937, 1.        ]])

Извлечем коэффициент r

In [ ]:
r_value = corr_matrix[0, 1]
r_value

np.float64(0.9343293664033213)

Определим направление связи

> r коэффициент равен 0.9343293664033213, положителен
>
> значит связь прямая, чем больше товаров, тем больше время

Силу линейной связи

> Сила очень велика, тк |r| > 0,90

Сформулируйте возможное объяснение наблюдаемой зависимости с точки зрения
работы склада и доставки.

> чем больше позиций тем больше сборщик ищет и собирает товары, тем дольше едет до заказчика

**Вопрос для анализа**

Можно ли на основании высокого значения корреляции утверждать:
«Увеличение количества товаров является причиной увеличения времени доставки»?
Объясните ответ.

> Это не может быть единственной причиной, просто это линейная логичная зависимость, но не единственная из возможных

## Задание 2.5. Корреляция и скрытые факторы

**Рассмотрим ситуацию:**

Летом одновременно увеличиваются:

• продажи мороженого;

• количество несчастных случаев на воде.

На основании этих данных аналитик делает вывод:

> «Продажа мороженого приводит к увеличению количества несчастных случаев на воде».

**Что необходимо сделать**
1. Объясните, почему высокая корреляция не доказывает наличие причинно-следственной
связи.
> Корреляция не означает прямую зависимость, это лишь синхронно изменяемые величины из-за общих фыакторов (лето, например)
2. Определите фактор, который может влиять одновременно на оба показателя.
>  Лето
3. Объясните механизм возникновения наблюдаемой корреляции.
> Лето - жарко, люди идут кушать мороженое, а ткаже эти же или иные люди идут на водоемы, и иногда возникают несчастные случаи, но мороженки не первопричина этих событий, хотя статистически они оба возрастают
4. Приведите собственный пример ложной корреляции из бизнеса, технологий, медиа или
повседневной жизни.
> Наступает сентябрь: появляется в жизни учеба и плохая погода. Рост числа покупок зонтиков коррелирует с ростом числа покупок тетрадок. это ложная корреляция. Если в июле пойдет дожджь никто не будет покупать тетрадки (в таком количестве)

# Блок 3. Bootstrap и статистические гипотезы

## Задание 3.1. Bootstrap и доверительный интервал

In [ ]:
x = np.array([28, 31, 29, 45, 32, 30])
x.mean()

np.float64(32.5)

In [ ]:
B = 5000
bootstrap_means = np.empty(B)

np.random.seed(42)
for i in range(B):
    bootstrap_sample = np.random.choice(x, size=len(x), replace=True)
    bootstrap_means[i] = bootstrap_sample.mean()

Доверительный интервал

In [ ]:
np.quantile(bootstrap_means, 0.025)

np.float64(29.166666666666668)

In [ ]:
np.quantile(bootstrap_means, 0.975)

np.float64(37.666666666666664)

Вопросы для анализа
1. Почему Bootstrap используется именно с выборкой с возвращением?
> Чтобы сымитировать выборку из генеральной совокупности. Без возвращения мы бы просто переставляли те же значения, и разброс был бы неверным
2. Как наличие наблюдения 45 влияет на Bootstrap-распределение среднего?
> Это выброс он сдвигает bootstrap-распределение вправо и расширяет интервал
3. Почему результаты Bootstrap могут заметно изменяться при использовании очень
маленькой исходной выборки?
> Чем меньше n, тем сильнее каждый элемент влияет на результат, тем шире интервал
4. Можно ли утверждать: *«Вероятность того, что истинное среднее находится внутри рассчитанного 95%-го доверительного интервала, равна 95%»*?
> Нет, это не вероятность для конкретного интервала. Если повторять эксперимент много раз, 95% построенных таким образом интервалов накроют истинное среднее

## Задание 3.2. Формулировка гипотез и ошибки I и II рода

Продуктовая команда внедрила новый алгоритм распределения заказов между курьерами.

Цель алгоритма — уменьшить среднее время доставки.

Обозначим:

• \mu_A — истинное среднее время доставки в контрольной группе;

• \mu_B — истинное среднее время доставки в группе с новым алгоритмом.

> Н0 - новый алгоритм распределения заказов не влияет на время доставки
>
> Н1 - новый алгоритм распределения заказов уменьшает среднее время доставки
>
> I рода ошибка ("ложная тревога") - алгоритм сочли эффективным хотя это не так, например, время доставки сократилось но из-за других факторов
>
> II рода ошибка ("пропущенный эффект") - алгоритм сочли неэффективным хотя он эффективен, например, время сократилось, но из-за непредвиденных обстаятельств относительно увеличилось, хотя в такой же ситуации старый алгоритм показал бы еще большее время

**Ситуация для анализа**

Полноценная эксплуатация нового алгоритма обходится компании в 5 млн рублей в год.

Какая ошибка может привести к прямым напрасным расходам компании — ошибка I или ошибка II рода?

Обоснуйте ответ

> К прямым напрасным расходам приведет ошибка I рода. Компания потратит бюджет на технологию, которая никак не улучшает показатели

## Задание 3.3. p-value и принятие решения

После проведения A/B-теста статистический критерий дал:

p-value = 0.001

Уровень значимости:

α = 0.05

**Что необходимо сделать**

1. Дайте определение p-value.
> это вероятность наблюдаемых или более крайних данных при условии что нулевая гипотеза верна

2. Объясните, почему утверждение:

«p-value = 0.001 означает, что вероятность истинности нулевой гипотезы равна 0.1%» является неверным.
> p-value рассматривается при верности Н0 гипотезы, и не является ее вероятностью. Следовательно утверждение неверно.

3. Определите, является ли результат статистически значимым при $\alpha = 0.05$.
> Результат статистически значим, так как p-value строго меньше альфа, те данные странные для этой гипотезы

**Ситуация для анализа**

Предположим, эксперимент проведён на 1 000 000 пользователей.

Новый алгоритм уменьшил среднее время доставки всего на 3 секунды.

При этом статистический критерий дал:

p-value = 0.001

Эксплуатация нового алгоритма стоит компании 5 млн рублей в год.

**Вопросы**

1. Является ли результат статистически значимым?
> результат значим, и Н0 отвергаем
2. Достаточно ли одного p-value для принятия решения о внедрении?
> нет, так как оно не отражает бизнес-значимость
3. Почему статистическая значимость не обязательно означает практическую или бизнесзначимость?
> потому что оно не отражает проактический эффект данных. система может приносить определенный результат, но не отражать экономической значимости, вложений может быть много и система может работать но может и не отбить вложения. Мы лишь подтвердили что эффект не случаен
4. Какие дополнительные показатели необходимо учитывать перед раскаткой алгоритма на всех пользователей?
> Окупаемость, размер эффекта системы,

**Итоговый продуктовый вывод**

Представьте, что вы отвечаете за принятие решения о внедрении алгоритма.

Сформулируйте итоговую рекомендацию и обоснуйте её с использованием трёх составляющих принятия решения:

1. Статистическая значимость — насколько убедительны данные с точки зрения статистического критерия.

2. Размер и практическая значимость эффекта — насколько изменение существенно для пользователя и продукта.

3. Экономическая целесообразность — превышает ли ожидаемый эффект затраты, риски и стоимость эксплуатации решения.

> 1. Статистически результат значим и эффект не случаен. Но это необходиоме, но недостаточное условие для внедрения.
>
> 2. Практически эффект ничтожен - 3 секудны не заметны для пользователя, а метрики заметно не изщменятся.
>
> 3. Экономически убыточно - ускорение доставки на 3 секунды не окупят ежегодные вложения и сомнительные side-эффекты.
>
> Необходимо оценить побочные эффекты внедрения (может они где-то улучшают еще какие-то показатели, а где-то ухудшают), попробовать снизить стоимоть эксплуатации, увеличить эффект

# Итоговый вопрос лабораторной работы

Представьте, что вы получили набор данных о работе интернет-магазина.

Среднее время доставки уменьшилось после изменения алгоритма, но одновременно:

• вырос разброс времени доставки;

• p90 практически не изменился;

• p95 увеличился;

• статистический тест показал p-value < 0.05.

Сформулируйте, какие дополнительные вопросы вы зададите перед тем, как рекомендовать внедрение изменения на всех пользователей.

В ответе отдельно рассмотрите:

• качество данных;

• размер эффекта;

• распределение результатов;

• выбросы;

• процентили;

• статистическую значимость;

• практическую значимость;

• бизнес-эффект.

> 1. Какой период рассматривался при замере показателей? Условия сравнения одинаковы? *(Качество данных)*
> 2. Насколько уменьшилось среднее? Какой размер эффекта считается практически значимым? *(Размер эффекта)*
> 4. Как изменилась форма, а не только среднее? *(Распределение)*
> 5. Это реальные выбросы или ошибки данных? *(Выбросы)*
> 6. Почему p90 не изменился, а p95 вырос? *(Процентили)*
> 7. Проверялась ли значимость изменений в хвостах? Рост p95 — статистически значим или случаен? *(Статистическая значимость)*
> 8. Заметит ли пользователь изменение?*(Практическая значимость)*
> 9. Сколько стоит внедрение и эксплуатация алгоритма? *(Бизнес-эффект)*
> 10. Окупается ли эффект затратами? *(Бизнес-эффект)*
> 11. Не вырастут ли затраты на компенсации за медленные доставки (промокоды, возвраты)? *(Бизнес-эффект)*
